Copy src and config file to solution while building

vmayoral opened the pull request:
This becomes useful when iterating over optimizations in the code or the configuration options. Having a solution-specific backup allows to get back to each change and generate reports on how changes affected performance.

benjmarshall committed:
Merge branch 'release-1.0.0.a5' into develop

Merge changes from release branch.

vmayoral committed:
Add generic except to avoid issues when files do not exist

vmayoral committed:
Add solution stats to status command

vmayoral committed:
Copy src and config file to solution while building

vmayoral committed:
Improve styles of status command

vmayoral committed:
Update status report

benjmarshall committed:
Modify file expection handling to be compactible with python 2 and python 3.

benjmarshall committed:
Merge branch 'vmayoral-aux' into develop

Merge changes from PR to fix file handling expections in python2.

benjmarshall commented on hlsclt/build_commands/build_commands.py:
```diff
@@ -43,11 +43,19 @@ def do_start_build_stuff(ctx):
     file.write("export_design -format ip_catalog" + "\n")
     file.write("export_design -format sysgen" + "\n")
+    # Copy the src/ files as well as the config file to keep track of the changes over solutions
+    import shutil
```
Can we move this to the top please to keep all the imports together?

vmayoral replied:
Addressed at https://github.com/benjmarshall/hlsclt/pull/12/commits/4f279abe517401a65221606f37b487c1848e26a5.

benjmarshall reviewed (commented):
The code block to backup solution settings has been added to the default build function which will only execute if the user has not specified a build sub-command. This function, along with the sub-command specific function execute early in the 'build' sequence as they simply add the correct line to a TCL script for Vivado to execute. I think this code block would be better placed in the callback function which executes after the vivado run has finished. This means we can be sure that the vivado operations have actually executed and correctly created the new solution, and allows the code to be executed for any valid combination for sub-commands. I think realistically we only need to backup the the solution settings if a new synthesis has happened, do you think this is correct, or should we backup for every call, even if only a simulation is being run?

vmayoral committed:
Implement --stats option in status command

vmayoral committed:
Re-organize code

Address requests from
https://github.com/benjmarshall/hlsclt/pull/12/files/a654e40078a7e24ddff55ccbb47a21256d0a76af

vmayoral committed:
Improve help message for -s option in status

vmayoral committed:
hlsclt status -s, fix bugs

benjmarshall committed:
Merge branch 'solutionstats' of https://github.com/vmayoral/hlsclt into vmayoral-solutionstats

Merge feature developed in PR to add solution stats output to status
command.

benjmarshall committed:
Merge branch 'vmayoral-solutionstats' into develop

Merge feature to display solution stats in status command output.

benjmarshall reviewed (changes requested):
Thanks for the changes. It looks like one of your lines of code has been accidentally left in the default build function. Line 47 where you get locally the solution number. Could you please remove this line, then I'm happy to merge in the new feature.

vmayoral committed:
build command, remove leftover

vmayoral commented:
@benjmarshall, done.

benjmarshall commented:
Thanks for the changes. I've done another quick test here today and found a couple of bugs which need resolved. 

Firstly if the build command is run without a -k option the hlsclt throws an error when attempting to overwrite the existing files:

`Traceback (most recent call last):
  File "/home/ben/anaconda3/bin/hlsclt", line 11, in <module>
    load_entry_point('hlsclt', 'console_scripts', 'hlsclt')()
  File "/home/ben/anaconda3/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/ben/anaconda3/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/ben/anaconda3/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/ben/anaconda3/lib/python3.6/site-packages/click/core.py", line 1093, in invoke
    return _process_result(rv)
  File "/home/ben/anaconda3/lib/python3.6/site-packages/click/core.py", line 1031, in _process_result
    **ctx.params)
  File "/home/ben/anaconda3/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/ben/anaconda3/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/mnt/centos_share/Vivado_Projects/hlsclt_vmayoral/hlsclt/build_commands/build_commands.py", line 161, in build_end_callback
    do_end_build_stuff(ctx,sub_command_returns,report)
  File "/mnt/centos_share/Vivado_Projects/hlsclt_vmayoral/hlsclt/build_commands/build_commands.py", line 119, in do_end_build_stuff
    shutil.copytree("src", destiny_src)
  File "/home/ben/anaconda3/lib/python3.6/shutil.py", line 315, in copytree
    os.makedirs(dst)
  File "/home/ben/anaconda3/lib/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: 'proj_simple_adder/solution1/src'
`
We need to add some handling for this, i suggest it should be safe enough to correctly remove and then copy fresh files.

Secondly if the build command is called for the first time on a fresh project with the -k option then the code creates a "solution2" folder instead of "solution1". This may be an existing bug (i'll try and do some further testing on the master branch) but it is made obvious by this new feature as any subsequent build commands then fail with the first bug, i.e. can't overwrite files, even if the -k option is provided.

benjmarshall reviewed (changes requested):
New changes requested, as per previous comment detailing bugs.

benjmarshall commented:
Underlying bug for the second case listed above has been found in master. Detailed in #14. A fix will be provided on a new branch which can be merged into this PR.

benjmarshall committed:
Fix solution numbering function bug where -k used on a fresh project created solution 2.

benjmarshall committed:
Merge branch 'fix_solution_numbering' of https://github.com/benjmarshall/hlsclt into backupsrc

Merge fix for solution numbering bug to allow further development of
sourc backup feature.

benjmarshall commented:
Merged in fix for solution numbering to allow easier development on this feature.

EDIT: I seem to have merged in all updates from develop, however I am going to leave this as it is because I think this PR will be merged back onto develop shortly anyway!

benjmarshall committed:
Fix error when trying to overwrite the backed up source directory of a pre-existing solution.

benjmarshall commented:
First issue described above fixed by latest commit. @vmayoral can you check you agree with the logic of simply overwriting the previous source directory backup?

vmayoral reviewed (commented):
Very ok with this change.

vmayoral commented:
_Edited_ my my fork to get something similar https://github.com/vmayoral/hlsclt/commit/4b74fe06912e55d81e5067585f05a79443cb59c8.

benjmarshall commented:
Thanks @vmayoral. This PR has now been merged into develop. I will be pushing a new release shortly which incorporates all the new features.
